java/ql/src/semmle/code/java/security/Random.qll

```diff
@@ -2,15 +2,164 @@ import java
+   * if applicable to this method's type and a constant bound is known.
+   */
+  Expr getUpperBoundExpr() { none() }
+
+  /**
+   * Holds if this source of random data may return bounded values (e.g. integers between 1 and 10).
+   * If it does not hold, it may return any value in the range of its result type (e.g., any possible integer).
+   */
+  predicate resultMayBeBounded() { none() }
+
+  /**
+   * Gets the result of this source of randomness: either the method access itself, or some argument
+   * in the case where it writes random data to that argument.
+   */
+  abstract Expr getOutput();
+}
+
+/**
+ * A method access calling a method declared on `java.util.Random`
+ * that returns random data or writes random data to an argument.
+ */
+class StdlibRandomSource extends RandomDataSource {
+  Method m;
+
+  StdlibRandomSource() {
+    m = this.getMethod() and
+    m.getDeclaringType() instanceof StdlibRandom
+  }
+
+  override int getLowerBound() {
+    // If this call is to `nextInt(int)`, the lower bound is zero.
+    m.hasName("nextInt") and
+    m.getNumberOfParameters() = 1 and
+    result = 0
+  }
+
+  override Expr getUpperBoundExpr() {
+    // If this call is to `nextInt(int)`, the upper bound is the first argument.
+    m.hasName("nextInt") and
+    m.getNumberOfParameters() = 1 and
+    result = this.getArgument(0)
   }
+
+  override predicate resultMayBeBounded() {
+    // `next` may be restricted by its `bits` argument,
+    // `nextBoolean` can't possibly be usefully bounded,
+    // `nextDouble` and `nextFloat` are between 0 and 1,
+    // `nextGaussian` is extremely unlikely to hit max values.
+    m.hasName(["next", "nextBoolean", "nextDouble", "nextFloat", "nextGaussian"])
+    or
+    m.hasName(["nextInt", "nextLong"]) and
+    m.getNumberOfParameters() = 1
+  }
+
+  override Expr getOutput() {
+    if m.hasName("getBytes") then result = this.getArgument(0) else result = this
+  }
+}
+
+/**
+ * A method access calling a method declared on `org.apache.commons.lang3.RandomUtils`
+ * that returns random data or writes random data to an argument.
+ */
+class ApacheCommonsRandomSource extends RandomDataSource {
+  Method m;
+
+  ApacheCommonsRandomSource() {
+    m = this.getMethod() and
+    m.getDeclaringType() instanceof ApacheRandomUtils
+  }
+
+  override Expr getLowerBoundExpr() {
+    // If this call is to `nextInt(int, int)` or `nextLong(long, long)`, the lower bound is the first argument.
+    m.hasName(["nextInt", "nextLong"]) and
+    m.getNumberOfParameters() = 2 and
+    result = this.getArgument(0)
+  }
+
+  override Expr getUpperBoundExpr() {
+    // If this call is to `nextInt(int, int)` or `nextLong(long, long)`, the upper bound is the second argument.
+    m.hasName(["nextInt", "nextLong"]) and
+    m.getNumberOfParameters() = 2 and
+    result = this.getArgument(1)
+  }
+
+  override predicate resultMayBeBounded() {
+    m.hasName(["nextDouble", "nextFloat"])
```

Comment:
`nextFloat()` and `nextDouble()` return 0 to `MAX_VALUE`; is this really bounded?

Reply:
Per https://docs.oracle.com/javase/8/docs/api/java/util/Random.html they return something in the range 0-1

Reply:
But this is for `org.apache.commons.lang3.RandomUtils` which does not extend `Random`.

Reply:
Ah sure, I hadn't noticed they deviated. For now I'll allow these because they're lower-bounded; separately we can improve ArithmeticUncontrolled.ql, the only query that cares about the difference, to distinguish RNG routines that risk overflow from those that risk underflow.